worldedit-core/src/main/java/com/sk89q/worldedit/regions/shape/ArbitraryShape.java

```diff
@@ -37,21 +39,27 @@ public abstract class ArbitraryShape {
     private final int cacheOffsetX;
     private final int cacheOffsetY;
     private final int cacheOffsetZ;
+    @SuppressWarnings("FieldCanBeLocal")
     private final int cacheSizeX;
     private final int cacheSizeY;
     private final int cacheSizeZ;
 
     /**
      * Cache for expression results.
      *
-     * <p>
-     * Cache entries:
-     * 0 = unknown
-     * -1 = outside
-     * 1 = inside
-     * </p>
+     * <p>Possible cache entries:
+     * <ul>
+     * <li>null = unknown</li>
+     * <li>Optional.empty() = outside</li>
+     * <li>Optional.of(material) = inside</li>
+     * </ul>
+     *
+     * <p>You're <i>technically</i> not *supposed* to set variables of type <code>Optional</code> to null, but if we did
+     * that, we'd need to use <code>Optional&lt;Optional&lt;BaseBlock&gt;&gt;</code> and pre-fill the cache with
+     * <code>Optional.empty()</code>, so we just use null here and rely on Java to pre-fill the array with
+     * <code>null</code>s for us.
      */
-    private final byte[] cache;
+    private final Optional<BaseBlock>[] cache;
```

Comment:
Has any performance testing of this been done? This might substantially increase the memory usage & the CPU time (due to memory allocations & indirection) of the caching system here, potentially outweighing the underlying calculations

Reply:
This is now an `Object[]`, resolving.